File: act col/03/Exercise01.java
// =================================================================
//
// File: Exercise01.java
// Authors: 
//             A01704696 Samuel Octavio Gonzalez Azpeitia
//             A01704052 Martín Adrian Noboa Monar   
// Description: This file contains the code to count the number of
//				even numbers within an array using Java's Threads.
//
// Copyright (c) 2023 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

public class Exercise01 extends Thread {
	private static final int SIZE = 100_000_000;
	private int array[], start, end, result;

	public Exercise01(int start, int end, int array[]) {
		// place your code here
        this.array = array;
        this.start = start;
        this.end = end;
		
	}

    public int getResult() {    
        return result;
    }

	public void run() {
		// place your code here
        result = 0;
        for (int i = start; i < end; i++) {
            if (array[i] % 2 == 0) {
                result++;
            }
        }
	}

	public static void main(String args[]) {
		int array[] = new int[SIZE];
		int result = 0;
		long startTime, stopTime;
		double elapsedTime = 0;
		int blockSize;
		Exercise01 threads[];

		Utils.fillArray(array);
		Utils.displayArray("array", array);

		// place your code here
        blockSize = SIZE / Utils.MAXTHREADS;
        threads = new Exercise01[Utils.MAXTHREADS];
		
		elapsedTime = 0;
		System.out.printf("Starting...\n");
		for (int i = 0; i < Utils.N; i++) {
			startTime = System.currentTimeMillis();

			// place your code here
            for (int j = 0; j < threads.length; j++) {
                threads[j] = new Exercise01(j * blockSize, (j + 1) * blockSize, array);
            }

            for (int j = 0; j < threads.length; j++) {
                threads[j].start();
            }

            for (int j = 0; j < threads.length; j++) {
                try {
                    threads[j].join();
                    result += threads[j].getResult();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }

			stopTime = System.currentTimeMillis();

			elapsedTime += (stopTime - startTime);
		}
        


		System.out.printf("result = %d\n", result);
		System.out.printf("avg time = %.5f ms\n", (elapsedTime / Utils.N));
	}
}

File: act col/06/Exercise01.cpp
// =================================================================
//
// File: exercise01.cpp
// Author(s):
//      Martin Adrian Noboa Monar - A01704052
//      Samuel Octavio González Azpeitia - A01704696
// Description: This file contains the code to count the number of
//				even numbers within an array using the TBB 
//				technology. To compile:
//				g++ exercise01.cpp -o app -I/usr/local/lib/tbb/include -L/usr/local/lib/tbb/lib/intel64/gcc4.4 -ltbb
//
// =================================================================

#include <iostream>
#include <iomanip>
#include <chrono>
#include <tbb/blocked_range.h>
#include <tbb/parallel_reduce.h>
#include "utils.h"

using namespace std;
using namespace std::chrono;
using namespace tbb;

// array size
#define SIZE 1000000000

// implement your code
class Exercise01{
private:
    double result;
    int *array;
public:
    Exercise01(int* a): array(a), result(0){};

    Exercise01(Exercise01 &other, split): array(other.array), result(0) {}

    double getResult() const {
        return result;
    }

    void operator() (const blocked_range<int> &r) {
        for (int i = r.begin(); i != r.end(); i++) {
            if (array[i] % 2 == 0) {
                result++;
            }
        }
    }

    void join(const Exercise01 &other) {
		result += other.result;
	}
};



int main(int argc, char* argv[]) {
	int *array;
    double result;
	// These variables are used to keep track of the execution time.
	high_resolution_clock::time_point start, end;
	double timeElapsed;

	array = new int[SIZE];
	fill_array(array, SIZE);
	display_array("array", array);

	cout << "Starting...\n";
	timeElapsed = 0;
	for (int j = 0; j < N; j++) {
		start = high_resolution_clock::now();

		// call the implemented function
        Exercise01 even(array);
        parallel_reduce(blocked_range<int>(0, SIZE), even);
		result = even.getResult();

		end = high_resolution_clock::now();
		timeElapsed += 
			duration<double, std::milli>(end - start).count();
	}
	cout << "result = " << result << "\n";
	cout << "avg time = " << fixed << setprecision(3) 
		 << (timeElapsed / N) <<  " ms\n";

	delete [] array;
	return 0;
}